crates/adapters/src/transport/pubsub/test.rs

```diff
@@ -0,0 +1,518 @@
+};
+use log::info;
+use pipeline_types::program_schema::Relation;
+use proptest::prelude::*;
+use serde::Serialize;
+
+static EMULATOR_PROJECT_ID: &str = "feldera-test";
+
+fn probe_port(address: &str) -> bool {
+    let connection_result =
+        TcpStream::connect_timeout(&address.parse().unwrap(), Duration::from_secs(1));
+
+    match connection_result {
+        Ok(_) => true, // Port is open
+        Err(e) => {
+            match e.kind() {
+                io::ErrorKind::ConnectionRefused | io::ErrorKind::TimedOut => false, // Port is closed or unreachable
+                _ => {
+                    panic!("Error probing port {}: {:?}", address, e);
+                }
+            }
+        }
+    }
+}
+
+#[cfg(feature = "pubsub-gcp-test")]
+async fn credentials_json_from_file() -> Result<Vec<u8>, anyhow::Error> {
+    let path = match std::env::var("GOOGLE_APPLICATION_CREDENTIALS") {
+        Ok(s) => Ok::<_, anyhow::Error>(std::path::Path::new(s.as_str()).to_path_buf()),
+        Err(_e) => {
+            // get well known file name
+            if cfg!(target_os = "windows") {
+                let app_data = std::env::var("APPDATA")?;
+                Ok(std::path::Path::new(app_data.as_str())
+                    .join("gcloud")
+                    .join("application_default_credentials.json"))
+            } else {
+                match home::home_dir() {
+                    Some(s) => Ok(s
+                        .join(".config")
+                        .join("gcloud")
+                        .join("application_default_credentials.json")),
+                    None => bail!("no home directory found"),
+                }
+            }
+        }
+    }?;
+
+    let credentials_json = tokio::fs::read(path).await?;
+
+    Ok(credentials_json)
+}
+
+struct TestPublisher {
+    client: Client,
+    publisher: Publisher,
+    topic: Topic,
+}
+
+impl TestPublisher {
+    fn with_emulator(topic: &str, emulator: &str) -> Self {
+        TOKIO.block_on(async {
+            let mut config = ClientConfig::default();
+            config.project_id = Some(EMULATOR_PROJECT_ID.to_string());
+            config.environment = Environment::Emulator(emulator.to_string());
+            let client = Client::new(config).await.unwrap();
+
+            // Create topic.
+            let topic = client.topic(topic);
+            if topic.exists(None).await.unwrap() {
+                topic.delete(None).await.unwrap();
+            }
+
+            topic.create(None, None).await.unwrap();
+
+            Self {
+                client,
+                publisher: topic.new_publisher(None),
+                topic,
+            }
+        })
+    }
+
+    fn with_auth(topic_id: &str) -> Self {
+        TOKIO.block_on(async {
+            info!("Creating test publisher using Application Default Credentials");
+
+            let config = ClientConfig::default().with_auth().await.unwrap();
+            //config.project_id = Some(PROJECT_ID.to_string());
```

Comment:
are these needed or can they be deleted?
maybe they are a form of documentation?